### DISCUSSION TITLE
Custom compailer api

### DISCUSSION THREAD

**Discussion by Akiyamka** [avajs/ava#2917] [Ideas]
Many developers use slightly different tech stack. JavaScript have probably the most varied
ecosystem.
jsx, vue, svelte, typescript or es6 modules, bundlers written on rust or go, and many more exotic variations require different approaches. Ava have support some of them, but it very custom and baked in ava codebase (like babel support).
It would be cool if ava expose some transform api, that can be used for pre-compiling assets before test run. Something similar already used in avajs/babel, but it limited by cjs, and looks very verbose.

Interesting approach I found in react-cosmos - they [just generate file](https://github.com/react-cosmos/react-cosmos/blob/b0fffb368d853378a95dfaffff2ee775cfcf99dd/packages/react-cosmos/src/plugins/userDepsFile.tsx) where imported all files that we want to compile and delegate build step of that file to user. It allow me compile files with vite instead of webpack without any limits, with all settings (like aliases and other transforms) out of box.

It possible to do something similar for avajs?

**Comment by novemberborn** [comment]
> Ava have support some of them, but it very custom and baked in ava codebase (like babel support).

It's actually pretty generic, which is why we were able to move the details into separate packages.  And there is value in having officially supported integrations, but I agree it would also be good to have a generic "here load this path and you can integrate" option.

How would you want to configure this?

```js
export default {
  providers: [
	'@my/vue-ava',
    ['@my/vue-svelte', { customConfig: true }],
  ],
}
```

Could have an import path (must be `import()` compatible) that exports a `default` that then negotiates the protocol, as in here:

https://github.com/avajs/typescript/blob/34da7bf008bd94e4844562ed86f5446faae6c087/index.js#L71

Or an array with an import path and the second argument being options that are then passed in.

What do you think?
